File: android/api-surface-check/src/negativeCheckNotifierValue/kotlin/jp/kamusoft/ksdialogs/apicheck/RejectsMismatchedNotifierValue.kt
package jp.kamusoft.ksdialogs.apicheck

import android.view.View
import jp.kamusoft.ksdialogs.DialogViewRegistry

/**
 * 宣言結果型と異なる値では完了報告できない (core/ADR-0003)。
 *
 * このソースは `-Pksdialogs.negativeCheck.notifierValue` を付けたときだけビルドに加わり、
 * **コンパイルエラーで失敗すること**が期待結果になる。
 * 期待する診断: Argument type mismatch: actual type is 'String', but 'Boolean' was expected.
 */
public object RejectsMismatchedNotifierValue {
    public fun register(registry: DialogViewRegistry) {
        registry.register(ConsumerDialogViewModel::class) { _, notifier ->
            notifier.complete("完了")
            View(this)
        }
    }
}

File: maui/KsDialogs.Maui.ApiSurfaceCheck/NegativeChecks/RejectsMismatchedNotifierValue.cs
using Microsoft.Maui.Controls;

namespace KsDialogs.ApiSurfaceCheck.NegativeChecks;

/// <summary>宣言結果型と異なる値では完了報告できない (core/ADR-0003)。</summary>
/// <remarks>
/// このソースはビルドプロパティ KsDialogsNegativeCheckNotifierValue を true にしたときだけ加わり、
/// <b>コンパイルエラーで失敗すること</b>が期待結果になる。
/// 期待する診断: CS1503 (string から bool へ変換できない)
/// </remarks>
internal static class RejectsMismatchedNotifierValue
{
    public static void Register(DialogViewRegistry registry)
    {
        registry.Register((ConsumerDialogViewModel _, DialogNotifier<bool> notifier) =>
        {
            notifier.Complete("bool ではない値");
            return new Label();
        });
    }
}
